File: packages/node/src/resources/tasks.ts
import type { MissionInbox } from '../client.js';
import type {
  ListTasksParams,
  Task,
  TaskList,
  TaskOutputs,
  TaskStatsSummary,
  WaitForTaskOptions,
} from '../types.js';
import { TERMINAL_TASK_STATUSES } from '../types.js';

/**
 * The `tasks` resource. Access via `mi.tasks`.
 *
 * Bulk operations (bulk create/verify/delete domains, etc.) run as
 * background tasks. Endpoints that dispatch a task return a `{ taskId }`,
 * and you use this resource to poll it to completion — or use
 * {@link Tasks.waitFor} which does the polling for you.
 */
export class Tasks {
  /** @internal */
  constructor(private readonly client: MissionInbox) {}

  /** List the account's tasks with optional filters. */
  async list(params: ListTasksParams = {}): Promise<TaskList> {
    const query: Record<string, unknown> = {};
    if (params.type !== undefined) query.type = params.type;
    if (params.status !== undefined) query.status = params.status;
    if (params.page !== undefined) query.page = params.page;
    if (params.limit !== undefined) query.limit = params.limit;
    return this.client.request<TaskList>({
      method: 'GET',
      path: '/api/tasks',
      query,
    });
  }

  /** Retrieve a task by id. */
  async get(id: string): Promise<Task> {
    return this.client.request<Task>({
      method: 'GET',
      path: `/api/tasks/${encodeURIComponent(id)}`,
    });
  }

  /** Cancel a pending or in-progress task. Already-terminal tasks return unchanged. */
  async cancel(id: string): Promise<Task> {
    return this.client.request<Task>({
      method: 'DELETE',
      path: `/api/tasks/${encodeURIComponent(id)}/cancel`,
    });
  }

  /**
   * Retrieve a task's execution log. Pass `since` (task-output id) to fetch
   * only entries created after a previous poll — useful for streaming a log.
   */
  async getOutputs(id: string, since?: string): Promise<TaskOutputs> {
    const query: Record<string, unknown> = {};
    if (since !== undefined) query.since = since;
    return this.client.request<TaskOutputs>({
      method: 'GET',
      path: `/api/tasks/${encodeURIComponent(id)}/outputs`,
      query,
    });
  }

  /** Account-wide task counts by status. */
  async getStatsSummary(): Promise<TaskStatsSummary> {
    return this.client.request<TaskStatsSummary>({
      method: 'GET',
      path: '/api/tasks/stats/summary',
    });
  }

  /**
   * Poll a task until it reaches a terminal state (`COMPLETED`, `FAILED`, or
   * `CANCELLED`), then resolve with the final task. Throws if the deadline
   * passes first.
   *
   * @example
   * ```ts
   * const { taskId } = await mi.domains.bulkCreate({
   *   domains: [{ domainName: 'a.com' }, { domainName: 'b.com' }],
   * });
   * const done = await mi.tasks.waitFor(taskId, {
   *   pollInterval: 3000,
   *   onProgress: (t) => console.log(t.progress),
   * });
   * console.log('done:', done.status, done.result);
   * ```
   */
  async waitFor(id: string, options: WaitForTaskOptions = {}): Promise<Task> {
    const pollInterval = options.pollInterval ?? 2000;
    const timeout = options.timeout ?? 300_000;
    const deadline = Date.now() + timeout;

    for (;;) {
      const task = await this.get(id);
      options.onProgress?.(task);
      if ((TERMINAL_TASK_STATUSES as readonly string[]).includes(task.status)) return task;
      if (Date.now() + pollInterval > deadline) {
        throw new Error(`Timed out after ${timeout}ms waiting for task ${id} (last status: ${task.status})`);
      }
      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }
  }
}

File: packages/python/src/missioninbox/resources/tasks.py
"""The `tasks` resource. Access via `mi.tasks`."""

from __future__ import annotations

import time
from typing import Any, Callable, Optional
from urllib.parse import quote

from .._http import HttpClient

TERMINAL_TASK_STATUSES: tuple[str, ...] = ("COMPLETED", "FAILED", "CANCELLED")


class Tasks:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        type: Optional[str] = None,
        status: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> dict[str, Any]:
        query: dict[str, Any] = {}
        if type is not None:
            query["type"] = type
        if status is not None:
            query["status"] = status
        if page is not None:
            query["page"] = page
        if limit is not None:
            query["limit"] = limit
        return self._http.request("GET", "/api/tasks", query=query)

    def get(self, id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/api/tasks/{quote(id, safe='')}")

    def cancel(self, id: str) -> dict[str, Any]:
        return self._http.request("DELETE", f"/api/tasks/{quote(id, safe='')}/cancel")

    def get_outputs(self, id: str, since: Optional[str] = None) -> dict[str, Any]:
        query = {"since": since} if since is not None else None
        return self._http.request(
            "GET",
            f"/api/tasks/{quote(id, safe='')}/outputs",
            query=query,
        )

    def get_stats_summary(self) -> dict[str, Any]:
        return self._http.request("GET", "/api/tasks/stats/summary")

    def wait_for(
        self,
        id: str,
        *,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
        on_progress: Optional[Callable[[dict[str, Any]], None]] = None,
    ) -> dict[str, Any]:
        """Poll a task until it reaches a terminal state.

        Args:
            id: Task id.
            poll_interval: Seconds between polls. Default 2.0.
            timeout: Give up after this many seconds. Default 300 (5 min).
            on_progress: Called with each poll result (useful for progress UIs).

        Raises:
            TimeoutError: When the deadline passes before completion.
        """
        deadline = time.monotonic() + timeout
        while True:
            task = self.get(id)
            if on_progress is not None:
                on_progress(task)
            status = task.get("status")
            if status in TERMINAL_TASK_STATUSES:
                return task
            if time.monotonic() + poll_interval > deadline:
                raise TimeoutError(
                    f"Timed out after {timeout}s waiting for task {id} (last status: {status})"
                )
            time.sleep(poll_interval)
